internal/repository/index.go

```diff
@@ -514,6 +583,43 @@ func (idx *Index) TreePacks() restic.IDs {
 	return idx.treePacks
 }
 
+// merge merges indexes, i.e.
+// idx.merge(idx2) merges the contents of idx2 into idx.
+func (idx *Index) merge(idx2 *Index) error {
+	idx.m.Lock()
+	defer idx.m.Unlock()
+
+	if !idx2.final {
+		return errors.New("index to merge is not final!")
+	}
+
+	// insert all index entries
+	idx2.blob.forEach(func(h restic.BlobHandle, entry indexEntry) {
+		for _, blob := range idx2.withDuplicates(h, entry) {
+			// packIndex is changed as idx2.pack is appended to idx.pack, see below
+			blob.packIndex += len(idx.packs)
+			if !idx.blob.setIfNotExist(h, blob) {
+				idx.duplicates[h] = append(idx.duplicates[h], blob)
+			}
+		}
+	})
+
+	idx.packs = append(idx.packs, idx2.packs...)
+	idx.treePacks = append(idx.treePacks, idx2.treePacks...)
+	idx.ids = append(idx.ids, idx2.ids...)
+	idx.supersedes = append(idx.supersedes, idx2.supersedes...)
+
+	// idx2 is no longer used
+	idx2.blob.clear()
```

Comment:
Clearing the second index ensures that my obscure race condition, becomes rather easy to trigger. In addition this also violates the locking policy used for indexes: Accessing or modifying parts of an index requires the user to hold the lock. That is `merge` must also call `idx2.m.Lock()` (I think we should be safe from circular lock requests...).

Reply:
I added the lock to idx2. Thanks for finding this!